core/primitives/src/validator_signer.rs

```diff
@@ -222,6 +237,13 @@ impl ValidatorSigner for InMemoryValidatorSigner {
         self.signer.sign(witness_bytes.as_slice())
     }
 
+    fn sign_partial_encoded_state_witness(
+        &self,
+        part: &PartialEncodedStateWitnessInner,
+    ) -> Signature {
+        self.signer.sign(&borsh::to_vec(part).unwrap())
```

Comment:
Something to keep in mind: anywhere we verify signatures on network messages, an attacker can attempt to overload the node by sending spam. That is because verification (even of invalid signatures) is computationally expensive. 

If you take a look [at chunk distribution](https://github.com/saketh-are/nearcore/blob/cbf977aef7685c705df6fe1c7bce142eadf7076a/chain/chunks/src/lib.rs#L1089-L1105), this problem is solved by building a merkle tree over the parts and signing the root. That way receiving nodes only need to verify and cache one signature to process all of the parts. After that they just verify the hashes of the individual parts.

Reply:
I remember us discussing that since these are messages from trusted chunk validators, we can assume most of them wouldn't be malicious?

In case we do get a malicious message, we can close the connection to that peer?

Between signature and merkle proof, now we would need to publish the merkle root for the proofs somewhere, probably the chunk header.

Reply:
On further thought yes, I think that should be ok. As you said we know the identity of the producer we expect to validate a signature from and if we ever get something invalid/unexpected we can drop the connection to the peer which sent us the offending message. I have been thinking about this from the context of signed network edges, where we find ourselves in the position of having to verify large batches of signatures made using arbitrary public keys. We don't really have the same problem here.